packages/@react-aria/dnd/src/DragManager.ts

```diff
@@ -0,0 +1,524 @@
+/*
+ * Copyright 2020 Adobe. All rights reserved.
+ * This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+import {announce} from '@react-aria/live-announcer';
+import {ariaHideOutside} from '@react-aria/overlays';
+import {DragEndEvent, DragItem, DropActivateEvent, DropEnterEvent, DropEvent, DropExitEvent, DropOperation, DropTarget as DroppableCollectionTarget} from '@react-types/shared';
+import {getDragModality, getTypes} from './utils';
+import {useEffect, useState} from 'react';
+
+let dropTargets = new Map<Element, DropTarget>();
+let dropItems = new Map<Element, DroppableItem>();
+let dragSession: DragSession = null;
+let subscriptions = new Set<() => void>();
+
+interface DropTarget {
+  element: HTMLElement,
+  getDropOperation?: (types: Set<string>, allowedOperations: DropOperation[]) => DropOperation,
+  onDropEnter?: (e: DropEnterEvent, dragTarget: DragTarget) => void,
+  onDropExit?: (e: DropExitEvent) => void,
+  onDropTargetEnter?: (target?: DroppableCollectionTarget) => void,
+  onDropActivate?: (e: DropActivateEvent, target?: DroppableCollectionTarget) => void,
+  onDrop?: (e: DropEvent, target?: DroppableCollectionTarget) => void,
+  onKeyDown?: (e: KeyboardEvent, dragTarget: DragTarget) => void
+}
+
+export function registerDropTarget(target: DropTarget) {
```

Comment:
I understand the appeal of a function to pass to useEffect or useLayoutEffect, but it seems easy to forget that this has cleanup and must be returned to useEffect
should we make a new standard name for these kinds of functions so we can lint for them? or can we put the useEffect/useLayoutEffect in here so it's impossible to forget?

Reply:
Ehh... maybe? This is only used in a couple places though.

Reply:
only used in a couple of places and missed once ;)